### ISSUE TITLE
[Suggestion] Alternative to cracking DRM

### ISSUE THREAD

**Issue by cheatfreak47** [qjimbo/NMSLegacyVersionInstaller#1]
We discussed this on the discord already but I'm placing my findings here so they don't get buried on discord, although feel free to [refer to that conversation](https://discord.com/channels/887444645632692274/887444645632692277/1255646377979936839) for additional context.

At the moment this installer, as part of it's process, involves cracking the game to run legacy builds of No Man's Sky, and this is more or less completely unnecessary.

<ins>**the %command% trick**</ins>
Steam has a section under each game where the user can edit the game's launch options, and a little known trick one can use in this launch options section is to shift the default launch commands forward by inserting the `%command%` shortcut in this launch options section. You can then place a _**full path in quotes**_ to another executable before it, which will make the game launch whatever executable you want with the default commands as arguments. The program it launches instead will behave and inherit the DRM friendly environment that No Man's Sky is provided by being launched from Steam this way, and thus, all child processes of any programs launched this way, also will inherit this environment.

I got this idea from SMAPI, the Stardew Modding API, which uses this same %command% trick to launch itself before the main game, allowing modding the game without needing hacky tricks like DLL injection.

Anyway, An extremely lazy way of using this would be to directly insert the path to the old No Man's Sky build you want to run here directly, and this does work, but I would suggest instead to insert a custom program that disregards the %command% argument, and can handle launching any and all versions as you want and accept it's own launch commands as well, and possibly take care of things like save redirection, and then launch the desired version.

I have a fully functioning implementation of this idea available [here](https://github.com/cheatfreak47/misc-scripts/blob/main/NMS-Redir.ahk). You can find some basic documentation on this implementation [here](https://github.com/cheatfreak47/misc-scripts?tab=readme-ov-file#nms-redirahk-autohotkey-11-script) as well as in the comments inside the script. 

My implementation is what I would consider to be a minimalist one, since it has no GUI and handles all functionality through command line arguments, but it should be possible to write as complex a program as you want, even a fullscreen GUI with controller support or something. Although, personally I'm more partial to the minimalist approach.

An example launch options string from my implementation:
```
"C:\Program Files (x86)\Steam\steamapps\common\No Man's Sky\NMS-Redir.exe" %command%
```

From there, if the user wants they can add additional data to this string, like `--version X.XX.X` or `--logging` or even specify a custom depots path using `--depotspath C:\Path\Goes\Here`. If the user doesn't provide any of these, it simply launches the live version of No Man's Sky- meaning, users launching the game through Steam itself won't notice anything different, it operates as a passthrough in that context. It also has tons of error handling and validation as well, since this implementation is actually based on very similar script I wrote for doing this same thing for old Terraria versions.

Anyway, It's also worth noting that, if taking advantage of the %command% trick, you can set up shortcuts on Windows to the game that is using the %command% trick that each specify additional arguments. `C:\Path\To\Steam.exe -applaunch 275850 --version 1.09.1` would launch the existing No Man's Sky install including it's %command% trick, but append `--version 1.09.1` to the launch, meaning it's possible to make your own Windows shortcuts that each provide additional commands. This is good, since it means if a proper installer were made, you can simply instruct the user to change their launch options on Steam *once* and they won't need to change it again unless they're disabling or removing the software, and just use Start Menu shortcuts to launch old versions or whatever.

All this also works well for Save Redirection and doesn't mess with Steam Cloud or anything like that, since Steam is launching your executable as if it is No Man's Sky, so you can perform whatever operations you want before the real game launches, and whatever operation you want after the real game closes, and Steam won't update Cloud Sync until our program closes, since it believes our program is part of No Man's Sky.

**Comment by qjimbo** [comment]
Thank you for the suggestion, I did start working on this with the SteamWrap project on my GitHub, however, decided to put it aside at the time. The added complexity of requiring the user to have steam running and rerouting the command just overall feels less practical than simply just getting to a point where the EXE can be double clicked. I apologize for taking a few years to address it fully.

While cracking the DRM is somewhat problematic, to date there have been no issues or complaints received, I believe because we have limited the scope to 10 year old versions of the game.

In a recent update to the installer, I now take advantage of our replacement of the Steam DLL to allow self-contained rerouting of the savegames by version by using spoofed SteamIDs too, which this approach would lose.

I may investigate other alternatives to modifying the EXE though as I agree that this would be cleaner.

**Comment by bmaupin** [comment]
> a little known trick one can use in this launch options section is to shift the default launch commands forward by inserting the %command% shortcut in this launch options section

I don't think that's necessary. Here's a much easier fixed that I just tested myself:

Make a file in the No Man's Sky directory named `steam_appid.txt` with this as the only contents:

```
275850
```

Or alternatively set the game launch options in Steam to:

`SteamAppId=275850 %command%`

**Comment by cheatfreak47** [comment]
Yeah that does seem to work! It seems this instructs the steamapi dll what "game" it should "attach" to steam with for the license
that's pretty neat!
